source/common/router/header_formatter.cc

```diff
@@ -2,13 +2,109 @@
 
 #include <string>
 
+#include "envoy/common/optional.h"
+
 #include "common/access_log/access_log_formatter.h"
+#include "common/common/utility.h"
+#include "common/config/metadata.h"
+#include "common/json/json_loader.h"
 
 #include "fmt/format.h"
 
 namespace Envoy {
 namespace Router {
 
+namespace {
+
+std::string formatUpstreamMetadataParseException(const std::string& params,
+                                                 const EnvoyException* cause = nullptr) {
+  std::string reason;
+  if (cause != nullptr) {
+    reason = fmt::format(", because {}", cause->what());
+  }
+
+  return fmt::format("Incorrect header configuration. Expected format "
+                     "UPSTREAM_METADATA([\"namespace\", \"k\", ...]), actual format "
+                     "UPSTREAM_METADATA{}{}",
+                     params, reason);
+}
+
+std::function<std::string(const Envoy::AccessLog::RequestInfo&)>
+parseUpstreamMetadataField(const std::string& params_str) {
+  if (params_str.empty() || params_str[0] != '(' || params_str[params_str.size() - 1] != ')') {
+    throw EnvoyException(formatUpstreamMetadataParseException(params_str));
+  }
+
+  std::vector<std::string> params;
+  try {
+    Json::ObjectSharedPtr parsed_params =
+        Json::Factory::loadFromString(StringUtil::subspan(params_str, 1, params_str.size() - 1));
+
+    for (const auto& param : parsed_params->asObjectArray()) {
+      params.emplace_back(param->asString());
+    }
+  } catch (Json::Exception& e) {
+    throw EnvoyException(formatUpstreamMetadataParseException(params_str, &e));
+  }
+
+  // Minimum parameters are a metadata namespace (e.g. "envoy.lb") and a metadata key.
+  if (params.size() < 2) {
+    throw EnvoyException(formatUpstreamMetadataParseException(params_str));
+  }
+
+  return [params](const Envoy::AccessLog::RequestInfo& request_info) -> std::string {
+    Upstream::HostDescriptionConstSharedPtr host = request_info.upstreamHost();
+    if (!host) {
+      return "";
```

Comment:
I slightly prefer "return std::string()" (same for other places in this function)